text/0085-api-keys.md

```diff
@@ -0,0 +1,463 @@
+
+Granular management of API keys is added to MeiliSearch. It is possible to restrict the access of an API key to certain actions on specific indexes.
+
+### II. Motivation
+
+To make MeiliSearch more reliable for teams, we extend the management and the possibilities of restrictions for the management of a MeiliSearch instance by introducing a concrete API resource (`API Key`). Security is a critical need, often tricky to negotiate as the stakes are high for a company.
+
+### III. Glossary
+
+| Term               | Definition |
+|--------------------|------------|
+| Main Key         | This is the main key that allows you to create other API keys. The main key is defined by the user when launching MeiliSearch, thus gives access to the `/keys` API endpoint. |
+| API Key            | API keys are stored and managed from the endpoint `/keys` by the main key holder. These are the keys used by the technical teams to interact with MeiliSearch at the level of the client code. |
+
+### IV. Personas
+
+| Persona | Role |
+|---------|------|
+| Anna    | Anna has an `ssh` access to the MeiliSearch instance and manages it on a daily basis. |
+| Mark    | Mark is a developer from the same company as Anna. He will implement the code to communicate with MeiliSearch to solve technical/product needs. |
+
+### V. `API Key` Explanations
+
+#### Summary Key Points
+
+- Occurrences of `master key` are changed by `main key`.
+- `MEILI_MASTER_KEY` environment variable is renamed `MEILI_MAIN_KEY`.
+- `--master-key` CLI option is renamed `--main-key`.
+- `/keys` management is restricted to the main key.
+- When a main key is set at MeiliSearch first-launch, we generate two pre-configured `API Key` resources. A `Default Search API Key` restricted to the search action and a `Default Admin API Key` to handle all operations (except managing API Keys) on MeiliSearch.
+- New endpoints are added to manage the `API Key` resource.
+- `API keys` can have restrictions on which methods can be accessed via an `actions` list; they can also `expiresAt` a specific date time and be restricted to a specific set of `indexes`.
+
+#### Managing `API Key`
+
+![](https://i.imgur.com/mAUFnNb.png)
+
+`Anna` manages the MeiliSearch instance; she uses the main key she defined at startup to generate an `API Key` resource for `Mark` to use in his client code to communicate with the MeiliSearch instance.
+
+`Anna` can define access rights to certain indexes to define an expiration date and also authorized `actions` for an `API Key` (See API Key Actions List Definition Part).
+
+Only the main key allows managing the API keys.
+
+#### Main Key
+
+The main key exists to secure a MeiliSearch instance. As soon as a main key is set via the  `MEILI_MAIN_KEY` environment variable or the `--main-key` CLI option , the endpoint `/keys` is accessible only for the main key holder. It can be seen as a super admin key; It must be shared only with people who have to manage the security of a MeiliSearch instance.
```

Comment:
The `keys` endpoints will be inaccessible since the user will not be able to fill in a valid `main` key if he has not generated one at the MeiliSearch launch. I'll fill this in! Thank you for pointing this.